src/core/graph.go

```diff
@@ -184,6 +216,66 @@ func (graph *BuildGraph) linkDependencies(fromTarget, toTarget *BuildTarget) {
 	}
 }
 
+// checkArchitectureCompatibility checks that two targets have compatible architectures. It panics if they don't.
+func checkArchitectureCompatibility(from, to *BuildTarget) {
+	if from.Label.Arch != "" && to.Label.Arch != "" && from.Label.Arch != to.Label.Arch {
+		panic(fmt.Sprintf("%s requires a dependency on %s, but it's not available for that architecture", from.Label, to.Label.toArch("")))
+	}
+}
+
+// cloneTargetForArch returns a build target for the given architecture. It's otherwise
+// identical to the original one.
+func (graph *BuildGraph) cloneTargetForArch(target *BuildTarget, arch string) *BuildTarget {
+	archLabel := target.Label.toArch(arch)
+	t, present := graph.targets[archLabel]
+	if present {
+		return t
+	}
+	graph.targets[archLabel] = target // Sentinel, so we don't recurse into here incorrectly.
```

Comment:
should there be a lock when interacting with `graph.targets`? Presumably it's fine without because even if different goroutines overwrote each other, the contents would still be equivalent?